Restructures the "Highlighting" documentation (#134439)
elastic/elasticsearch: docs/reference/elasticsearch/rest-apis/how-es-highlighters-work-internally.md

---
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/highlighting.html
applies_to:
  stack: all
---

# How highlighters work internally [how-es-highlighters-work-internally]

Given a query and a text (the content of a document field), the goal of a highlighter is to find the best text fragments for the query, and highlight the query terms in the found fragments. For this, a highlighter needs to address several questions:

> * How to break a text into fragments?

Reviewer: Could we maybe have these link to the relevant subsection?

* How to find the best fragments among all fragments?
* How to highlight the query terms in a fragment?

> ## How to break a text into fragments? [_how_to_break_a_text_into_fragments]

Reviewer 2: ```suggestion
## How to break a text into fragments [_how_to_break_a_text_into_fragments]
```

Relevant settings: `fragment_size`, `fragmenter`, `type` of highlighter, `boundary_chars`, `boundary_max_scan`, `boundary_scanner`, `boundary_scanner_locale`.

Plain highlighter begins with analyzing the text using the given analyzer, and creating a token stream from it. Plain highlighter uses a very simple algorithm to break the token stream into fragments. It loops through terms in the token stream, and every time the current term’s end_offset exceeds `fragment_size` multiplied by the number of created fragments, a new fragment is created. A little more computation is done with using `span` fragmenter to avoid breaking up text between highlighted terms. But overall, since the breaking is done only by `fragment_size`, some fragments can be quite odd, e.g. beginning with a punctuation mark.

Unified or FVH highlighters do a better job of breaking up a text into fragments by utilizing Java’s `BreakIterator`. This ensures that a fragment is a valid sentence as long as `fragment_size` allows for this.

> ## How to find the best fragments? [_how_to_find_the_best_fragments]

Reviewer 2: ```suggestion
## How to find the best fragments [_how_to_find_the_best_fragments]
```

FVH doesn’t need to analyze the text and build an in-memory index, as it uses pre-indexed document term vectors, and finds among them terms that correspond to the query. FVH scores each fragment by the number of query terms found in this fragment. Similarly to plain highlighter, score of individual term is equal to its boost value. In contrast to plain highlighter, all query terms are counted, not only unique terms.

Unified highlighter can use pre-indexed term vectors or pre-indexed terms offsets, if they are available. Otherwise, similar to Plain Highlighter, it has to create an in-memory index from the text. Unified highlighter uses the BM25 scoring model to score fragments.

> ## How to highlight the query terms in a fragment? [_how_to_highlight_the_query_terms_in_a_fragment]

Reviewer 2: ```suggestion
## How to highlight the query terms in a fragment [_how_to_highlight_the_query_terms_in_a_fragment]
```

Relevant settings:  `pre-tags`, `post-tags`.

The goal is to highlight only those terms that participated in generating the *hit* on the document. For some complex boolean queries, this is still work in progress, as highlighters don’t reflect the boolean logic of a query and only extract leaf (terms, phrases, prefix etc) queries.

Plain highlighter given the token stream and the original text, recomposes the original text to highlight only terms from the token stream that are contained in the low-level match information structure from the previous step.

FVH and unified highlighter use intermediate data structures to represent fragments in some raw form, and then populate them with actual text.

A highlighter uses `pre-tags`, `post-tags` to encode highlighted terms.

> ## An example of the work of the unified highlighter [_an_example_of_the_work_of_the_unified_highlighter]

Reviewer 2: ```suggestion
## How the unified highlighter works [_an_example_of_the_work_of_the_unified_highlighter]
```

Reviewer 2: Consider also a few more subheadings in this section for scannability

> Let’s look in more details how unified highlighter works.

Reviewer 2: ```suggestion
In this section, we'll explore how the unified highlighter works in more detail with an example.
```

There's a lot of language cleanup like these we could do here, but I'll leave that to your discretion.
